docs/study-def-tricks.md

```diff
@@ -0,0 +1,318 @@
+	    ),
+        }
+    variables = var_signature(f"{name}_1", codelist, index_date, return_expectations)
+    for i in range(2, n+1):
+        variables.update(var_signature(f"{name}_{i}", codelist, f"{name}_{i-1} + 1 day", return_expectations))
+    return variables
+```
+
+This can then be used inside a study definition for example as follows:
+
+
+```py
+study = StudyDefinition(
+
+	...
+	
+	**with_these_clinical_events_date_X(
+		name = "probable_covid_date", 
+		codelist = probable_codes, 
+		returning="date",
+		on_or_after = "index_date",
+		n = 5, 
+		return_expectations = {
+			"date": {"earliest": "2020-05-01", "latest": "2021-06-01"},
+			"rate": "uniform",
+			"incidence": 0.05,
+		},
+	),
+	
+	...
+)
+```
+
+This can be used for any variable where you want to return information for a series of consecutive events. 
+
+The will create `n` columns in the dataset. A few things to be aware of:
+
+* This does not make for efficient SQL queries. It is creating a new query for each event, rather than a query for each set of events. This will slow down extraction time considerably.
+* The dummy data won't necessarily appear in date order. It may be possible to tinker with the function to make this work. Another workaround is to reorder the columns post-extraction, for example here: https://github.com/opensafely/covid-vaccine-effectiveness-research/blob/be747894e1fadf525e391c01477a8ac532613b42/analysis/R/data_process.R#L229 
+* It's likely that you'll want to to convert this wide-formatted data to long format, 
+
+## Better dummy data for categorical variables
+For categorical variables with a small number of categories, the standard way of creating dummy data is easy to use, for example:
+
+```py
+sex=patients.sex(
+        return_expectations={
+            "rate": "universal",
+            "category": {"ratios": {"M": 0.49, "F": 0.51}},
+        }
+    )
+```
+
+However, if the number of categories is large, for instance small geographical regions like MSOAs, then it's a pain to have to write out all these by hand. An alternative is to create an external file containing the category levels, import this as a python dict, and pass this dict to the `return_expectations` argument. For example:
+
+```py
+import pandas as pd
+
+STPs = pd.read_csv(
+    filepath_or_buffer = './lib/STPs.csv'
+)
+
+dict_stp = { stp : 1/len(STPs.index) for stp in STPs['stp_id'].tolist() }
+
+study = StudyDefinition(
+ 
+    ...
+
+    stp = patients.registered_practice_as_of(
+        index_date,
+        returning="stp_code",
+        return_expectations={
+            "category": {"ratios": dict_stp},
+        },
+    ),
+    
+    ....
+
+)
+
+```
+
+If the category levels are named with a predictable structure for example (`["stage 1", "stage 2", ...]`, or [`100, 200, 300, ...`]) then a dict comprehension can be used to create the levels, without having to import them. For example for IMD ranks rounded to the nearest 100:
```

Comment:
link to dict comprehensions in python docs?

Reply:
Sticking with official python documentation, I could link to [dictionaries](https://docs.python.org/3/tutorial/datastructures.html#dictionaries) or [list comprehensions](https://docs.python.org/3/tutorial/datastructures.html#list-comprehensions), but not dict comprehensions. So I've linked to both -- hopefully that's not a terrible idea